### PR TITLE
feat: use same server info for all map modes

### PR DESCRIPTION
**Pull request by kohlisid**
Fixes https://github.com/numaproj/numaflow/issues/1838

### CODE REVIEW

File: pkg/mapstreamer/server.go [hunk 1]
```diff
@@ -35,9 +36,16 @@ func (m *server) Start(ctx context.Context) error {
...
 
-	// write server info to the file
+	// write server info to the file, we need to add metadata to ensure selection of the
+	// correct map mode, in this case streaming map
+	serverInfo := info.GetDefaultServerInfo()
+	serverInfo.Metadata = map[string]string{info.MapModeMetadata: string(info.StreamMap)}
+	if err := info.Write(serverInfo, info.WithServerInfoFilePath(m.opts.serverInfoFilePath)); err != nil {
+		return err
+	}
+
```
Comment (KeranYang) [inline pkg/mapstreamer/server.go:46]: I see that you moved this piece of code from `PrepareServer`, now we have two places writing the server file, outside and inside of PrepareServer.

Can we choose one? either all within PrepareServer or none.

File: pkg/batchmapper/examples/batchmap-flatmap/pipeline.yaml [hunk 2]
```diff
@@ -8,15 +8,12 @@ spec:
       source:
         http: {}
     - name: go-split
-      metadata:
-        annotations:
-          numaflow.numaproj.io/batch-map: "true"
```
Comment (KeranYang) [inline pkg/batchmapper/examples/batchmap-flatmap/pipeline.yaml:13]: why do we change this back to map without changing the folder name batchmap?

File: pkg/info/types.go [hunk 3]
```diff
@@ -15,6 +15,17 @@ const (
...
+type MapMode string
+
+const (
+	UnaryMap  MapMode = "unary-map"
+	StreamMap MapMode = "stream-map"
+	BatchMap  MapMode = "batch-map"
+)
+
+// MapModeMetadata field is used to indicate which map mode is enabled
+const MapModeMetadata = "MAP_MODE"
```
Comment (KeranYang) [inline pkg/info/types.go:27]:
```suggestion
// MapModeKey is the key used in the server info metadata map to indicate which map mode is enabled. 
const MapModeKey = "MAP_MODE"
```

**Comment by kohlisid** [comment]
Map Stream with new map mode metadata

no annotations in spec
```
    - name: flatmap
      limits:
        readBatchSize: 1
      udf:
        container:
          image: quay.io/kohlisid/numaflow-go/map-flatmap-stream:stable
```

Map stream server with `mapstream.sock`

```
 kubectl logs simple-pl-go-flatmap-0-w5nqq -f
2024/07/16 21:28:04 starting the gRPC server with unix domain socket... /var/run/numaflow/mapstream.sock
```

Server info file  - `/var/run/numaflow/mapper-server-info`
```
kubectl logs simple-pl-go-flatmap-0-w5nqq numa -f | grep mapper
2024/07/16 21:27:42 Server info file /var/run/numaflow/mapper-server-info is not ready...


 kubectl logs simple-pl-go-flatmap-0-w5nqq numa -f | grep Info  
2024/07/16 21:28:05 ServerInfo: &{uds go 1.2.0-rc4 v0.7.0-rc2 map[MAP_MODE:stream-map]}


kubectl logs simple-pl-go-flatmap-0-w5nqq numa -f | grep enabled
2024-07-16T21:28:05.294120902Z  INFO    numaflow.MapUDF-processor       udf/map_udf.go:150 
Map mode enabled: Stream Map    {"pipeline": "simple-pl-go", "vertex": "flatmap", "protocol": "uds-grpc-map-udf"}

```

File: pkg/mapstreamer/server.go (hunk 1 above)
Comment (KeranYang) [reply pkg/mapstreamer/server.go:46]: I prefer moving back to PrepareServer by adding the ServerInfo object to the input parameters of PrepareServer's.

File: pkg/batchmapper/examples/batchmap-flatmap/pipeline.yaml (hunk 2 above)
Comment (KeranYang) [reply pkg/batchmapper/examples/batchmap-flatmap/pipeline.yaml:13]: Oh, we are no longer using annotations. Resolving...

Comment (KeranYang) [reply pkg/batchmapper/examples/batchmap-flatmap/pipeline.yaml:13]: sorry, still this pipeline seems not related to batch

**Comment by kohlisid** [comment]
Map Stream with new map mode metadata

no annotations in spec
```
    - name: go-split
      scale:
        min: 1
      udf:
        container:
          image: quay.io/kohlisid/numaflow-go/batch-map-flatmap:stable1
          imagePullPolicy: Always
```

Batch Map stream server with `batchmap.sock`

```
kubectl logs flatmap-go-split-0-0xcsa -f
2024/07/17 18:50:23 starting the gRPC server with unix domain socket... /var/run/numaflow/batchmap.sock

```

Server info file  - `/var/run/numaflow/mapper-server-info`
```
kubectl logs flatmap-go-split-0-0xcsa numa -f | grep mapper 
2024/07/17 18:50:08 Server info file /var/run/numaflow/mapper-server-info is not ready...



 kubectl logs flatmap-go-split-0-0xcsa numa -f | grep ServerInfo:
2024/07/17 18:50:24 ServerInfo: &{uds go 1.2.0-rc4 v0.8.0 map[MAP_MODE:batch-map]}


 kubectl logs flatmap-go-split-0-0xcsa numa -f | grep enabled    
{"level":"info","ts":"2024-07-17T18:50:24.456998519Z","logger":"numaflow.MapUDF-processor","caller":"udf/map_udf.go:173","msg":"Map mode enabled: Batch Map","pipeline":"flatmap","vertex":"go-split","protocol":"uds-grpc-map-udf"}


```

**Comment by kohlisid** [comment]
Unary Map with new map mode metadata

no annotations in spec
```
    - name: go-split
      scale:
        min: 1
      udf:
        container:
          image: quay.io/kohlisid/numaflow-go/map-flatmap:stable
          imagePullPolicy: Always
```

Batch Map stream server with `map.sock`

```
kubectl logs flatmap-go-split-0-mvcdg -f
2024/07/17 18:57:42 starting the gRPC server with unix domain socket... /var/run/numaflow/map.sock
```

Server info file  - `/var/run/numaflow/mapper-server-info`
```
kubectl logs flatmap-go-split-0-mvcdg -f numa | grep mapper
2024/07/17 18:57:27 Server info file /var/run/numaflow/mapper-server-info is not ready...




 kubectl logs flatmap-go-split-0-mvcdg -f numa | grep ServerInfo
2024/07/17 18:57:43 ServerInfo: &{uds go 1.2.0-rc4 v0.7.0-rc2 map[MAP_MODE:unary-map]}


 kubectl logs flatmap-go-split-0-mvcdg -f numa | grep enabled   
{"level":"info","ts":"2024-07-17T18:57:43.308422104Z","logger":"numaflow.MapUDF-processor","caller":"udf/map_udf.go:196","msg":"Map mode enabled: Unary Map","pipeline":"flatmap","vertex":"go-split","protocol":"uds-grpc-map-udf"}



```

**Comment by kohlisid** [comment]
Unary Map with no metadata (and new numaflow)

no annotations in spec
```
    - name: go-split
      scale:
        min: 1
      udf:
        container:
          image: quay.io/numaio/numaflow-go/map-flatmap:stable
          imagePullPolicy: Always
```

Unary Map stream server with `map.sock`

```
 kubectl logs flatmap-go-split-0-elw3l -f
2024/07/17 19:01:19 starting the gRPC server with unix domain socket... /var/run/numaflow/map.sock
```

Server info file  - `/var/run/numaflow/mapper-server-info`
```
 kubectl logs flatmap-go-split-0-elw3l -f numa | grep mapper
2024/07/17 19:01:06 Server info file /var/run/numaflow/mapper-server-info is not ready...




 kubectl logs flatmap-go-split-0-elw3l -f numa | grep ServerInfo
2024/07/17 19:01:19 **ServerInfo: &{uds go 1.2.0-rc4 v0.7.0-rc2 map[]**}



 kubectl logs flatmap-go-split-0-elw3l -f numa | grep enabled   
{"level":"info","ts":"2024-07-17T19:01:19.426399748Z","logger":"numaflow.MapUDF-processor","caller":"udf/map_udf.go:196","msg":"Map mode enabled: Unary Map","pipeline":"flatmap","vertex":"go-split","protocol":"uds-grpc-map-udf"}

```

File: pkg/info/server_info.go [hunk 4]
```diff
@@ -109,3 +109,11 @@ func Read(opts ...Option) (*ServerInfo, error) {
 	}
 	return info, nil
 }
+
+// GetDefaultServerInfo returns a ServerInfo object with the default fields populated for Go-SDK
+func GetDefaultServerInfo() *ServerInfo {
+	serverInfo := &ServerInfo{Protocol: UDS, Language: Go, MinimumNumaflowVersion: MinimumNumaflowVersion,
+		Version: GetSDKVersion()}
```
Comment (KeranYang) [inline pkg/info/server_info.go:116]: Now that we move GetSDKVersion to inside of GetDefaultServerInfo, can we make GetSDKVersion pkg private?

**Review by KeranYang — APPROVED** [review]

**Review by ayildirim21 — APPROVED** [review]

**Pull request merged**